Add SpanExporter.ForceFlush. (#1467)
open-telemetry/opentelemetry-specification: specification/trace/sdk.md

#### ForceFlush()

This is a hint to ensure that any tasks associated with `Spans` for which the
`SpanProcessor` had already received events prior to the call to `ForceFlush` SHOULD
be completed as soon as possible, preferably before returning from this method.

In particular, if the `SpanProcessor` has any associated exporters, it MUST call

> the exporter's `Export` with all spans for which this was not already done and

Reviewer: I think this shouldn't be "MUST". For example, if the processor has 1000 items, and it is calling the `exporter.Export` with a maximum size of 100 items, on the 200 items timeout limit is hit - the `processor.ForceFlush` should return as early as possible with an indication of timeout rather than push more items to the exporter.

Author: Interesting point, we could add wording on the timeout here. But I don't know how close we want to specify that. For example: It might make sense to heuristically determine in the SpanProcessor "Oh no, we are coming close to the timeout, we took average x seconds per batch but still have y batches left and only z seconds of the timeout. Better call exporter.ForceFlush now and only then Export more spans if time is still left".

Probably that wouldn't make sense though because if Export takes so long that we run into the timeout, ForceFlush would maybe be a no-op -- but it might also not be, e.g. `write` for a filebased exporter can take long, and still needs a `flush` at the end.

Specifying the details is complicated here, so maybe we should say:

> `ForceFlush` MUST be implemented such that these conditions are fulfilled:
>
> * If there was no timeout, `ForceFlush`MUST have been called after all spans were passed to `Export` of the Exporter.
> * If there was a timeout, at least one `Export` or `ForceFlush` operation was attempted.
> * ForceFlush MAY be called more than once if at least one Export call is in-between. Note: SpanProcessors might want to heuristically determine that they could run into a timeout and in response call `ForceFlush` before finishing export.
>
> If `ForceFlush` does not support a timeout, or an infinite timeout was given, then a stricter condition is required: The operation MUST call the exporter's `Export` with all spans for which this was not already done and invoke ForceFlush on the export afterwards exactly once.

Author: We could additionally specify that the timeout expressed as a deadline timepoint SHOULD be passed on to the exporter, i.e. if I call processor.ForceFlush(timeout=5) at T=2, then the deadline is D = T + timeout = 7, so the first export call may get timeout=5, and if it took 1, then the next export call invoked at T=3 will get timeout=4 (= D - T).

Reviewer: The "most permissive but maybe still useful spec" looks reasonable to me 👍

Reviewer 2: > I think this shouldn't be "MUST". For example, if the processor has 1000 items, and it is calling the exporter.Export with a maximum size of 100 items, on the 200 items timeout limit is hit - the processor.ForceFlush should return as early as possible with an indication of timeout rather than push more items to the exporter.

I feel we are too picky, and try to over specify things. Maybe as simple as "In particular, if the `SpanProcessor` has any associated exporters, it MUST call the exporter's `Export` with all spans for which this was not already done and then invoke `ForceFlush` on it. " If a timeout is given then the operation should return when deadline is hit.

If users calls with such a small timeout they get half of the work done. We should not try to ensure at least one export and one flush is called. If timeout is 1us then maybe only an export can be called, etc. So just explain the order of execution and describe that timeout should be interrupt this.

Author: >  We should not try to ensure at least one export and one flush is called

I agree, that's why I suggested "or". But I also feel that my suggestion in the comment above is overly complicated.

Author: I added 04e3e29.

then invoke `ForceFlush` on it.

`ForceFlush` SHOULD provide a way to let the caller know whether it succeeded,
failed or timed out.

`ForceFlush` SHOULD only be called in cases where it is absolutely necessary,
such as when using some FaaS providers that may suspend the process after an
invocation, but before the `SpanProcessor` exports the completed spans.

> `ForceFlush` SHOULD complete or abort within some timeout. `ForceFlush` can be

Reviewer 3: ```suggestion
`ForceFlush` SHOULD complete or abort within some timeout. `ForceFlush` MAY be
```

Author: I think this should be done for all the ForceFlush definitions and also Shutdown at once, it would be confusing if TracerProvider/SpanProcessor/Exporter differed only there.

#### ForceFlush()

This is a hint to ensure that any tasks associated with `Spans` for which the
`SpanProcessor` had already received events prior to the call to `ForceFlush` SHOULD
be completed as soon as possible, preferably before returning from this method.

In particular, if any `SpanProcessor` has any associated exporter, it SHOULD
try to call the exporter's `Export` with all spans for which this was not
already done and then invoke `ForceFlush` on it.

> The [built-in SpanProcessors](#built-in-span-processors) MUST do so.

Reviewer 3: @Oberon00 Please add an entry in the compliance matrix for that, otherwise it might be overlooked easily.

and the destination is unavailable). OpenTelemetry client authors can decide if they
want to make the shutdown timeout configurable.

> #### `ForceFlush()`

Reviewer 2: I think we can say, in case of stable releases a "no-op" implementation can be offered as a default. I think that is the best we can do :)

> This is a hint to ensure that any `Spans` passed to the exporter prior to the

Author: ```suggestion
This is a hint to ensure that the export of any `Spans` passed to the exporter prior to the
```

> This is a hint to ensure that the export of any `Spans` passed to the exporter prior to the

Reviewer: This is a bit vague, if another thread is calling `exporter.Export(batch)`, what do we expect `exporter.ForceFlush` to do?

Author: This is up to the exporter I think. But if Export and ForceFlush are called concurrently, then by definition we have a race condition, even if the Exporter does locking internally, so we cannot possibly guarantee that any Spans in that Export call are exported or not. The most we could offer would be a "all or none Spans of each concurrent Export call are affected" gurantee, but I don't think that would be helpful.

Reviewer: I guess we could change it from "any `Spans` passed to the exporter" to "any `Spans` that the exporter has received"?

Author: ```suggestion
This is a hint to ensure that the export of any `Spans`  the exporter has received prior to the
```

> call to `ForceFlush` SHOULD be completed as soon as possible, preferably before

Reviewer 3: ```suggestion
call to `ForceFlush` SHOULD be exported as soon as possible, preferably before
```
to avoid confusion with `end`ing spans

Author: I will do this instead: https://github.com/open-telemetry/opentelemetry-specification/pull/1467/files#r583035984

returning from this method.

`ForceFlush` SHOULD provide a way to let the caller know whether it succeeded,
failed or timed out.

`ForceFlush` SHOULD only be called in cases where it is absolutely necessary,
such as when using some FaaS providers that may suspend the process after an

> invocation, but before the `Processor` exports the completed spans.

Reviewer 3: ```suggestion
invocation, but before the exporter exports the completed spans.
```

> `ForceFlush` SHOULD complete or abort within some timeout. `ForceFlush` can be

Reviewer 3: ```suggestion
`ForceFlush` SHOULD complete or abort within some timeout. `ForceFlush` MAY be
```

Author: Same as https://github.com/open-telemetry/opentelemetry-specification/pull/1467/files#r583035024

and the destination is unavailable). OpenTelemetry client authors can decide if they
want to make the shutdown timeout configurable.

#### `ForceFlush()`

This is a hint to ensure that the export of any `Spans` the exporter has received prior to the
call to `ForceFlush` SHOULD be completed as soon as possible, preferably before
returning from this method.

`ForceFlush` SHOULD provide a way to let the caller know whether it succeeded,
failed or timed out.

`ForceFlush` SHOULD only be called in cases where it is absolutely necessary,
such as when using some FaaS providers that may suspend the process after an
invocation, but before the exporter exports the completed spans.

`ForceFlush` SHOULD complete or abort within some timeout. `ForceFlush` can be
implemented as a blocking API or an asynchronous API which notifies the caller
via a callback or an event. OpenTelemetry client authors can decide if they want to

> make the flush timeout configurable.

Reviewer 4: For blocking API, I think it is very hard to abort within some timeout. Like the implementation complies to the spec if it just does abort its entry point without doing any work. Should we say `ForceFlush` should return (complete or abort) as soon as possible if the given `timeout` is reached?

Author: This is copy & past from shutdown and the other ForceFlush. Please open an issue if you want tor reword this (should be consistent everywhere).
